packages/ra-auth-auth0/src/authProvider.ts

```diff
@@ -0,0 +1,153 @@
+    } = {
+        redirectOnCheckAuth: true,
+    }
+): AuthProvider => ({
+    // Used when the redirection to Auth0 is done from a custom login page
+    async login() {
+        client.loginWithRedirect({
+            authorizationParams: {
+                redirect_uri:
+                    loginRedirectUri ??
+                    `${window.location.origin}/auth-callback`,
+            },
+        });
+    },
+    // called when the user clicks on the logout button
+    async logout() {
+        const isAuthenticated = await client.isAuthenticated();
+        if (isAuthenticated) {
+            // need to check for this as react-admin calls logout in case checkAuth failed
+            return client.logout({
+                logoutParams: {
+                    returnTo: logoutRedirectUri || window.location.origin,
+                },
+            });
+        }
+    },
+    // called when the API returns an error
+    async checkError({ status }) {
+        if (status === 401 || status === 403) {
+            throw new Error('Unauthorized');
+        }
+    },
+    // called when the user navigates to a new location, to check for authentication
+    async checkAuth() {
+        const isAuthenticated = await client.isAuthenticated();
+        if (isAuthenticated) {
+            return;
+        }
+
+        if (redirectOnCheckAuth) {
+            localStorage.setItem(
+                PreviousLocationStorageKey,
+                window.location.href
+            );
+            client.loginWithRedirect({
+                authorizationParams: {
+                    redirect_uri:
+                        loginRedirectUri ??
+                        `${window.location.origin}/auth-callback`,
+                },
+            });
+        }
+    },
+    // called when the user navigates to a new location, to check for permissions / roles
+    async getPermissions() {
+        if (!(await client.isAuthenticated())) {
+            return;
+        }
+
+        // If Auth0 instance contains rules for returning permissions, use them
+        const claims = await client.getIdTokenClaims();
+        const roleProperty = Object.keys(claims).find(key =>
+            key.includes('role')
+        );
+        return claims[roleProperty];
+    },
+    async getIdentity() {
+        if (await client.isAuthenticated()) {
+            const user = await client.getUser();
+            return {
+                id: user.email,
+                fullName: user.name,
+                avatar: user.picture,
+            };
+        }
+        throw new Error('Failed to get identity.');
+    },
+    async handleCallback() {
+        const query = window.location.search;
+        if (query.includes('code=') && query.includes('state=')) {
+            try {
+                await client.handleRedirectCallback();
+                return;
+            } catch (error) {
+                console.log('error', error);
+                throw error;
+            }
```

Comment:
Error handling here is broken
It triggers an infinite loop unless we set `logoutOnFailure=false`.

Temporary (dirty) workaround:

```suggestion
            } catch (error) {
                console.log('error', error);
                // eslint-disable-next-line no-throw-literal
                throw { message: error.message, logoutOnFailure: false };
            }
```

It probably comes from one of the `queryClient.clear();` in [useLogout](https://github.com/marmelab/react-admin/blob/next/packages/ra-core/src/auth/useLogout.ts#L69).